storage/aptosdb/src/pruner/state_store/mod.rs

```diff
@@ -0,0 +1,265 @@
+        }))
+    }
+
+    fn least_readable_version(&self) -> Version {
+        self.least_readable_version.load(Ordering::Relaxed)
+    }
+
+    fn set_target_version(&self, target_version: Version) {
+        self.target_version.store(target_version, Ordering::Relaxed);
+    }
+
+    fn target_version(&self) -> Version {
+        self.target_version.load(Ordering::Relaxed)
+    }
+
+    fn record_progress(&self, least_readable_version: Version) {
+        self.least_readable_version
+            .store(least_readable_version, Ordering::Relaxed);
+        DIEM_STATE_STORAGE_PRUNER_LEAST_READABLE_VERSION.set(least_readable_version as i64);
+    }
+
+    fn is_pruning_pending(&self) -> bool {
+        self.least_readable_version() >= self.target_version()
+    }
+}
+
+impl StateStorePruner {
+    pub fn new(
+        db: Arc<DB>,
+        index_min_nonpurged_version: Version,
+        index_purged_at: Instant,
+    ) -> Self {
+        StateStorePruner {
+            db,
+            index_min_nonpurged_version: AtomicVersion::new(index_min_nonpurged_version),
+            index_purged_at: Mutex::new(index_purged_at),
+            target_version: AtomicVersion::new(0),
+            least_readable_version: AtomicVersion::new(0),
+        }
+    }
+
+    /// Purge the stale node index so that after restart not too much already pruned stuff is dealt
+    /// with again (although no harm is done deleting those then non-existent things.)
+    ///
+    /// We issue (range) deletes on the index only periodically instead of after every pruning batch
+    /// to avoid sending too many deletions to the DB, which takes disk space and slows it down.
+    fn maybe_purge_index(&self) -> anyhow::Result<()> {
+        const MIN_INTERVAL: Duration = Duration::from_secs(10);
+        const MIN_VERSIONS: u64 = 60000;
+
+        // A deletion is issued at most once in one minute and when the pruner has progressed by at
+        // least 60000 versions (assuming the pruner deletes as slow as 1000 versions per second,
+        // this imposes at most one minute of work in vain after restarting.)
+        let now = Instant::now();
+        if now - *self.index_purged_at.lock() > MIN_INTERVAL
+            && self.least_readable_version.load(Ordering::Relaxed)
+                - self.index_min_nonpurged_version()
+                + 1
+                > MIN_VERSIONS
+        {
+            let new_min_non_purged_version =
+                self.least_readable_version.load(Ordering::Relaxed) + 1;
+            self.db.range_delete::<StaleNodeIndexSchema, Version>(
+                &self.index_min_nonpurged_version(),
+                &new_min_non_purged_version, // end is exclusive
+            )?;
+            self.index_min_nonpurged_version
+                .store(new_min_non_purged_version, Ordering::Relaxed);
+            *self.index_purged_at.lock() = now;
+        }
+        Ok(())
+    }
+    pub fn index_min_nonpurged_version(&self) -> Version {
```

Comment:
nit: blank line

Reply:
I actually thought `cargo x fmt` takes care of these. Do you know if that's not the case or I am missing something?

Reply:
https://github.com/rust-lang/rustfmt/issues/1410
https://github.com/rust-dev-tools/fmt-rfcs/issues/57

Too long, I didn't read :P 
I take it as it doesn't take care of this case.